# Assess genome assemblies with Merqury

This session is part of [**Biodiversity Genomics Academy 2023**](https://BGA23.org)

## Session Leader(s)

[Arang Rhie](https://genomeinformatics.github.io/people/rhie/)
Staff Scientist, Genome Informatics Section, National Institutes of Health

Assisted by
[Steven Solar](https://genomeinformatics.github.io/people/solar/)
Bioinformatics Scientist, Genome Informatics Section, National Institutes of Health

[Brandon Pickett](https://genomeinformatics.github.io/people/pickett/)
Postdoctoral Fellow, Genome Informatics Section, National Institutes of Health

## Description

By the end of this session you will be able to:
1. Run Meryl and Merqury yourself
2. Learn how to interpret k-mer histogram and copy-number spectrum plots of haploid, diploid, or poly-ploid genome assemblies
3. Understand QV, completeness, haplotype switch outputs
4. Get familiar with some other useful tools in Meryl or Merqury

## Prerequisites

Basic understanding for genome assembly related terminologies, such as QV, N50, NG50, contigs, and scaffolds

!!! warning "Please make sure you MEET THE PREREQUISITES and READ THE DESCRIPTION above"

    You will get the most out of this session if you meet the prerequisites above.

    Please also read the description carefully to see if this session is relevant to you.
    
    If you don't meet the prerequisites or change your mind based on the description or are no longer available at the session time, please email tol-training at sanger.ac.uk to cancel your slot so that someone else on the waitlist might attend.
